### PR TITLE
fix(core): inline non-component $refs to avoid broken named imports (#398)

### LINKED ISSUE #398: References with tokens that contains placeholders cannot be found
## What are the steps to reproduce this issue?
Try to generate code for this schema.
```yaml
openapi: 3.0.3
info: {}
paths:
  /{id}:
    get:
      parameters:
        - in: path
          name: id
          schema:
            type: string
      responses:
        200:
          content:
            application/json:
              schema:
                $ref: "#/paths/~1%7Bid%7D/get/parameters/0/schema"
```

## What happens?
Orval fails because the reference cannot be resolved.

## What were you expecting to happen?
The reference to be resolved and code to be generated.

## Any logs, error output, etc?
```
🛑  <name> - Oups... 🍻. Ref not found: #/paths/~1%7Bid%7D/get/parameters/0/schema
```

## Any other comments?
You get this kind of reference when [bundling](https://apitools.dev/json-schema-ref-parser/docs/ref-parser.html#bundleschema-options-callback) a schema. Maybe the problem occurs with escaped characters in general.

## What versions are you using?

**Package Version:** 6.7.1

### PR DESCRIPTION
**Pull request by wadakatu**
## Summary

Fixes #398.

A `$ref` like `#/paths/~1%7Bid%7D/get/parameters/0/schema` (URL-encoded JSON Pointer to an inline schema, as emitted by `@apidevtools/json-schema-ref-parser` `bundle()`) used to crash orval with `Oups... 🍻. Ref not found`. That crash got fixed incidentally by `decodeURIComponent` handling, but the bug just changed shape — orval emitted

```ts
import type { Schema } from './model';
export type getByIdResponse200 = { data: Schema; status: 200 };
```

for an undeclared `Schema` type, producing TypeScript that no longer compiles. The user's report was therefore still unresolved on `master` (verified at `01536ba`).

### Root cause

`resolveValue` (`packages/core/src/resolvers/value.ts`) treated every resolved `$ref` as a named, exportable type. But only refs targeting `components.{schemas,responses,parameters,requestBodies}` actually have a corresponding `export type` emitted by `schema-definition.ts`. Refs to any other location — typically bundler artefacts pointing into `paths.*` — had no target import, leaving the dangling reference.

The four eligible component sections were also implicit in the codebase (only the keys of `RefComponentSuffix` declared them, with no comment explaining their special role).

### Fix

- Introduce `NAMED_COMPONENT_SECTIONS` in `packages/core/src/getters/ref.ts` as a single source of truth, with a doc comment explaining what makes these sections special. `RefComponent` type, `RefComponentSuffix`, and the new `COMPONENT_REF_PATTERN` all derive from it, so future maintainers update one place.
- Add `isComponentRef(ref: string): boolean` derived from that pattern.
- In `resolveValue`, when the user-facing `$ref` isn't a component ref, fall through to `getScalar` so the resolved schema is inlined exactly like a non-`$ref` schema would be — no broken named import is emitted.

### Before / after

The exact YAML from the issue:

```yaml
paths:
  /{id}:
    get:
      parameters:
        - in: path
          name: id
          schema: { type: string }
      responses:
        '200':
          content:
            application/json:
              schema:
                $ref: "#/paths/~1%7Bid%7D/get/parameters/0/schema"
```

Before:
```ts
import type { Schema } from './model';        // ← Schema is undeclared
export const getById = (...): Promise<AxiosResponse<Schema>> => { ... };
```

After:
```ts
export const getById = (...): Promise<AxiosResponse<string>> => { ... };
```

## Test plan

Three layers of coverage added:

- [x] **Unit** — `packages/core/src/getters/ref.test.ts` covers `isComponentRef` (named-component matrix, non-component cases including the issue's URL-encoded ref, JSON-Pointer-encoded names).
- [x] **Behavior** — `packages/core/src/resolvers/value.test.ts` exercises both branches of `resolveValue` (named import preserved for component refs, inline path for path-based refs).
- [x] **Snapshot regression** — `tests/specifications/issue-398-encoded-path-ref.yaml` (the exact YAML from the issue) wired into `tests/configs/default.config.ts`. Snapshot baseline committed.

Local CI gate (all passing):

- [x] `bun run format:check`
- [x] `bun run build`
- [x] `bun run typecheck`
- [x] `bun run lint`
- [x] `bun run test` — 1713/1713 pass
- [x] `bun run test:snapshots` — 3960/3960 pass
- [x] `bun run --filter orval-tests build` — `tsc` against generated TS for all 15 client modes

## Summary by CodeRabbit

* **Bug Fixes**
  * Inline path-based schema $refs correctly without emitting dangling imports, fixing issue #398.

* **Tests**
  * Added comprehensive tests for detecting named component refs and for value resolution (named vs path-based refs, encoded paths, and recursion guards).

* **Chores**
  * Added a repro OpenAPI spec, config and generated snapshot to validate encoded-path ref handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/orval-labs/orval/pull/3355)

### CODE REVIEW

File: packages/core/src/resolvers/value.ts [hunk 1]
```diff
@@ -25,6 +26,7 @@ export function resolveValue({
   formDataContext,
 }: ResolveValueOptions): ResolverValue {
   if (isReference(schema)) {
+    const refValue = (schema as { $ref?: string }).$ref;
```
[the remark this answers is not in the document]
Comment (wadakatu) [reply packages/core/src/resolvers/value.ts]: Good catch — fixed in 4edfe64aa. `OpenApiReferenceObject` is intersected with `{ $ref?: string }` (see types.ts:1498), so the cast was redundant. Kept the truthiness check because `$ref` is optional, but `schema.$ref` is now accessed directly.

File: packages/core/src/resolvers/value.ts [hunk 2]
```diff
@@ -33,6 +35,20 @@ export function resolveValue({
...
+    // See issue #398.
+    if (refValue && !isComponentRef(refValue)) {
+      const scalar = getScalar({
+        item: schemaObject,
+        name,
+        context,
+        formDataContext,
+      });
+      return { ...scalar, originalSchema: schemaObject, isRef: false };
+    }
```
[the remark this answers is not in the document]
Comment (wadakatu) [reply packages/core/src/resolvers/value.ts:71]: Addressed in 4edfe64aa. The inline branch now threads `refValue` into `context.parents` and short-circuits to `unknown` if the same ref is already on the stack. Anonymous recursive types can't be expressed in TypeScript anyway, so `unknown` is the safest break. Added a `breaks cycles on self-referential path-based refs` test in `value.test.ts`.

File: packages/core/src/getters/ref.ts [hunk 3]
```diff
@@ -17,6 +24,19 @@ export const RefComponentSuffix: Record<RefComponent, string> = {
...
+  String.raw`^#\/components\/(${NAMED_COMPONENT_SECTIONS.join('|')})\/[^/]+$`,
+);
+
+/**
+ * True iff `ref` targets a named slot eligible for emission as a TypeScript
+ * import. Used by `resolveValue` to decide between named import vs inlining
+ * the resolved schema.
+ */
+export function isComponentRef(ref: string): boolean {
+  return COMPONENT_REF_PATTERN.test(ref);
```
[the remark this answers is not in the document]
Comment (wadakatu) [reply packages/core/src/getters/ref.ts:37]: Declining this one. The regex is correct for all valid OpenAPI `$ref` strings:

- **`~1` JSON-Pointer escape**: works because `~1` itself contains no literal `/` — your test on line 40 confirms it. Not "by accident" — JSON Pointer (RFC 6901) defines `~1` as the only escape for a literal `/` inside a name segment, so a single segment after `/components/<section>/` will never contain a literal `/` if it's well-formed.
- **`%2F` URL-encoded slash**: this isn't a valid JSON-Pointer encoding for `/` — JSON Pointer uses `~1`, not URL percent-encoding. Tools that emit `%2F` are non-conformant. The current `getRefInfo` (`ref.ts:38`) does `decodeURIComponent` defensively, so even if such a malformed ref slipped in, the predicate accepting it actually matches the existing tolerant behavior.
- **`#/components/schemas/Foo` followed by another fragment**: per RFC 3986 a URI has at most one fragment identifier. `#/components/schemas/Foo#bar` is syntactically invalid, no conformant bundler emits it.

Splitting on `/` would be equivalent in semantics and slightly more verbose; not worth the change for parity.

File: tests/__snapshots__/default/issue-398-encoded-path-ref/endpoints.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * OpenAPI spec version: 1.0.0
+ */
+import axios from 'axios';
+import type { AxiosRequestConfig, AxiosResponse } from 'axios';
+
+export const getById = (
+  id: string,
+  options?: AxiosRequestConfig,
+): Promise<AxiosResponse<string>> => {
+  return axios.get(`/${id}`, {
```
[the remark this answers is not in the document]
Comment (wadakatu) [reply tests/__snapshots__/default/issue-398-encoded-path-ref/endpoints.ts:14]: The dangling-import case is already covered at three layers, so additional assertions would be redundant:

1. **Unit** (`value.test.ts`): `expect(result.imports).toHaveLength(0)` directly asserts no import is emitted for the path-ref case.
2. **Integration snapshot** (this file): the `endpoints.ts` snapshot is the canonical regression check — any future regression that re-introduces `Schema` here will fail `bun run test:snapshots`.
3. **Compile-level** (`tests/scripts/typecheck-generated.mjs`, run via `bun run --filter orval-tests build`): `tsc` runs against every generated client mode. A dangling import would fail this gate hard, which is the strongest possible compile-level guarantee.

Adding a fourth "compile validation in unit tests" layer would duplicate (3) at much higher complexity. The PR description summarises the three layers.

**Review by melloware — APPROVED** [review]
